resources/fetchers/kube_provider.go

```diff
@@ -62,19 +65,43 @@ func GetKubeData(watchers []kubernetes.Watcher) []fetching.Resource {
 	return ret
 }
 
-func (r K8sResource) GetID() (string, error) {
+func (r K8sResource) GetData() interface{} {
+	return r.Data
+}
+
+func (r K8sResource) GetMetadata() fetching.ResourceMetadata {
+	k8sObjMeta := r.GetK8sObjectMeta()
+	resourceID := k8sObjMeta.UID
+	resourceName := k8sObjMeta.Name
+
+	return fetching.ResourceMetadata{
+		ID:      string(resourceID),
+		Type:    k8sObjType,
+		SubType: r.GetSubType(),
+		Name:    resourceName,
+	}
+}
+
+func (r K8sResource) GetK8sObjectMeta() metav1.ObjectMeta {
 	k8sObj := reflect.Indirect(reflect.ValueOf(r.Data))
 	metadata, ok := k8sObj.FieldByName(k8sObjMetadataField).Interface().(metav1.ObjectMeta)
 	if !ok {
-		return "", fmt.Errorf("failed to retrieve object metadata")
+		logp.L().Errorf("failed to retrieve object metadata, Resource: %#v", r)
+		return metav1.ObjectMeta{}
 	}
 
-	uid := metadata.UID
-	return string(uid), nil
+	return metadata
 }
 
-func (r K8sResource) GetData() interface{} {
-	return r.Data
+func (r K8sResource) GetSubType() string {
+	k8sObj := reflect.Indirect(reflect.ValueOf(r.Data))
```

Comment:
try and redesign to avoid calling `reflect.Indirect(reflect.ValueOf(r.Data))` twice